OperationFuture: convert from polling to scheduled execution

vam-google opened the pull request:
https://github.com/googleapis/gax-java/issues/147
https://github.com/googleapis/gax-java/issues/146

Note, it was decided to not cancel server operation regardless of any state of OperationFuture. Doing so, might lead to breaking server running operation by an external poorly written client code, which did not necessarily starte the operation it tries to cancel.

vam-google commented on gax/src/main/java/com/google/api/gax/retrying/TimedRetryAlgorithm.java:
```diff
@@ -70,5 +72,13 @@
+   * @param nextAttemptSettings attempt settings, which will be used for the next attempt, if
+   *     accepted
+   */
+  boolean shouldCancel(TimedAttemptSettings nextAttemptSettings);
```
The retry algorithm is responsible for making a decision about what to do with the current retriable call. There are 3 possible outcomes for each call (basically a synonym of Future result in this context): 
- **succeded** (result accepted, `Future#get(`) will return result, `Future#isCancelled()` will be `false`); 
- **failed** (exception thorown on a last retry, `Future#get()` will throw `ExecutionException`, `Future#isCancelled()` will be `false`); 
- **canceled** (`Future#get()` will throw `CancellationException`, `Future#isCancelled()` will be `true`)

Just inverting shouldRetry does not allow to achieve this distinction between possible outcomes. Note how all possible combinations of two boolean values are already assigned in the following truth table, and there is no combination left to represent cancellation case:

shouldRetry | response (0) or exception(1) | retriable future
------------ | ------------- | -------------
0 | 0 | **success**
0 | 1 | **failed**, we gave up retrying
1 | 0 | **undefined**, we hope it succeeds next time, should be retried
1 | 1 | **undefined**, response returned, but not the one we want, should be retried
? | ? | **canceled**, for example when server returns response (not an error), saying that server operation was canceled, or it is sill an undefined case, but too much time has passed, so the polling should be canceled

Why do you think that a concept from the outer code is pushed here? I believe that cancel case was just missed in the original implementation, so adding it here.

vam-google replied:
Added the text in all 3 places (timed, result, and aggregating algos).

vam-google commented on gax/src/main/java/com/google/api/gax/retrying/RetryingFutureImpl.java:
```diff
@@ -83,10 +85,12 @@ public boolean cancel(boolean mayInterruptIfRunning) {
     synchronized (lock) {
       if (attemptFutureCallback != null) {
         if (attemptFutureCallback.attemptFuture.cancel(mayInterruptIfRunning)) {
+          cancelationInterruptStatus = mayInterruptIfRunning;
```
- corrected spelling. 
- i could not find a better name for it

vam-google commented on gax/src/main/java/com/google/api/gax/retrying/RetryAlgorithm.java:
```diff
@@ -34,28 +34,36 @@
+   * timedAlgorithm}.
    *
    * @param timedAlgorithm timed algorithm to use
+   * @param timedAlgorithm timed algorithm to use
```
Fixed

vam-google commented on gax/src/main/java/com/google/api/gax/retrying/RetryAlgorithm.java:
```diff
@@ -68,33 +76,75 @@ public TimedAttemptSettings createFirstAttempt() {
+  }
+
+  /**
+   * Returns {@code true} if another attempt should be made, or {@code false} otherwise.
```
Sorry, too much copy&paste. Corrected

vam-google commented on gax/src/main/java/com/google/api/gax/retrying/ResponseRetryAlgorithm.java:
```diff
@@ -0,0 +1,74 @@
+ * @param <ResponseT> response type
+ */
+@BetaApi
+public interface ResponseRetryAlgorithm<ResponseT> {
```
Agree, I had a feeling that something was wrong with two interfaces and could not figure out what exactly, now I now what - they must me one interface =). Thanks.

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationPollingCallable.java:
```diff
@@ -0,0 +1,119 @@
+/**
+ * A future callable, which relies on callback chaining to execute server polling (asking for a
+ * status of a specific operation: in progress, completed, canceled etc.). This implementation is
+ * called from {@link AttemptCallable}, essentially using for polling same logic which is used for
```
Corrected

vam-google commented:
Addressed review feedback. PTAL.

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationFuture.java:
```diff
@@ -29,348 +29,58 @@
 @BetaApi
-public final class OperationFuture<ResponseT extends Message> extends AbstractApiFuture<ResponseT> {
-  static final Duration DEFAULT_POLLING_INTERVAL = Duration.ofSeconds(1);
+public final class OperationFuture<ResponseT extends Message> implements ApiFuture<ResponseT> {
```
Why is there no way to get the operation name or metadata?

vam-google replied:
There is a way:
`operationFuture.getInitialFuture().get().getMetadata()` for metadata, and `operationFuture.getInitialFuture().get().getName()` for name.

I can add the wrapping methods, though they will be blocking methods. It feels not intuitive for a user: 1) why the methods are blocking and 2) (more importantly) on what they are blocking, and when a user can expect those to unblock. To answer the second quesition a user needs to know about existence of the initial future, so I decided just give him the instance explicitly, so he is not confused and also has more power (like adding a listener to it).

garrettjonesgoogle replied:
You're missing something important: the metadata returned by the API can change over time - some metadata types have progress % for the operation. So, you actually need to issue a new call each time to get updated metadata.

vam-google replied:
It is not different from how it was before: https://github.com/googleapis/gax-java/blob/master/gax-grpc/src/main/java/com/google/api/gax/grpc/OperationFuture.java#L305

So, I I don't miss anything, it is either old implementation was wrong too, or metadata does not change, or there are two different things here: initial call metadata and "ongoing" metadata.

garrettjonesgoogle replied:
Ahh, you're right; the old implementation was wrong too. I learned of the updating metadata after the implementation but didn't correct it.

I want the surface to allow a correct implementation later without people having to change their code. That means not getting the metadata from the initial call.

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationFuture.java:
```diff
@@ -29,348 +29,58 @@
-    public void wait(Duration duration) throws InterruptedException {
-      Thread.sleep(duration.toMillis());
-    }
+  public ApiFuture<Operation> getInitialFuture() {
```
Is this necessary? Does the class even need to hold the initial future?

vam-google replied:
Please check the previous comment https://github.com/googleapis/gax-java/pull/315#discussion_r119444944

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationPollingCallable.java:
```diff
@@ -0,0 +1,125 @@
+import java.util.concurrent.ExecutionException;
+
+/**
+ * A future callable, which relies on callback chaining to execute server polling (asking for a
```
no comma before "which"

vam-google replied:
Corrected. In Russian and Ukrainian there is no strict order of words in a sentence. You can say "I do" and "do I" and both of them can be either a question or a statement depending on the context and punctuation marks. As results there are much more commas in those languages, and that is why I keep putting so many commas in English =).

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationPollingCallable.java:
```diff
@@ -0,0 +1,125 @@
+ *
+ * @param <RequestT> type of the request
+ */
+class OperationPollingCallable<RequestT> implements FutureCallable<RequestT, Operation> {
```
This callable doesn't seem to do any polling - it just does a single call to check on the current state of the operation; I suggest a different name (I have no suggestions yet).

vam-google replied:
It does the polling, not by having an explicit loop inside, but by returning "correct" futures according to conditions. 

(I do not claim the following description to be completely correct or rigorous, but I believe it represent a valid point of view on this case):

This class is closer to functional programming approach than it is usual in java. I don't mean it is actually functional, especially since this class has state, but overall it is still somewhat close in essence. 

Specifically it does not have a procedural "while" loop for polling or anything like that which states how to perform the polling. 

Instead, this class contains two stateless functional classes `OperationRetryAlgorithm` and `OperationTimedAlgorithm` (aka first-class functions (or sets of functions), like in functional programming), which, when consumed by the "outer world" (`com.google.api.gax.retrying` package and `RetryingCallable`) make retrying to perform polling. 

I.e. one can put functions (methods in functional classes), declared in this class as arguments to the "retrying function", which will transform it into polling. The actual polling is done by retrying logic.

garrettjonesgoogle replied:
So it's only when the components are combined does this create a polling algorithm. The callable class here still seems like an `OperationCheckingCallable` to me. I'm thinking of when I look at the stack trace and I see the callables in it: it would be misleading to see this callable have "polling" in the name when it doesn't do that. 

As for the inner classes: it could be a little difficult to find them when looking at other code, because they are used in an unqualified way (e.g. from `OperationCallable`). I think it would be better for them to be in the same "area" as the other retry algorithms, which is top-level classes. (Both times I have done this review, it took me extra work to find them.)

I see what you're getting at with respect to the combination of the classes together creating a polling algorithm, but you still need other components to stitch it together to create the polling; these classes here aren't sufficient to do polling in isolation.

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationPollingCallable.java:
```diff
@@ -0,0 +1,125 @@
+    }
+  }
+
+  public static class OperationRetryAlgorithm implements ResultRetryAlgorithm<Operation> {
```
The two static classes aren't touched by the other code in this file - I suggest lifting them up or putting them somewhere else.

vam-google replied:
Please check the comment https://github.com/googleapis/gax-java/pull/315#discussion_r119438258

That comment answers both questions: why the class called polling and why those classes should be in that particular class.

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/AttemptCallable.java:
```diff
@@ -0,0 +1,108 @@
+    return null;
+  }
+
+  protected CallContext getNextCallContext(CallContext oldContext, Duration rpcTimeout) {
```
Why is this `protected`?

vam-google replied:
Mostly historical reasons. It was essential to an earlier version of the code, but then I changed it. Specifically what now is OperationPollingCallable used to inherit AttemptCallable overriding this method.

After changing it, I still kept the method protected, as it is doing some weird manipulations on call context, which are potentially not desirable by its children (since any non-final class can potentially have children, and automatically agrees to have children, otherwise it would be declared final).

The class itself is package-private, so it and its type hierarchy do not affect surface. I can change it back to private, if it matters.

garrettjonesgoogle replied:
It's easier to expand access than restrict it - let's start with private, and we can make it protected later if we need to.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/retrying/RetryAlgorithm.java:
```diff
@@ -68,33 +70,64 @@ public TimedAttemptSettings createFirstAttempt() {
-      Throwable prevThrowable, TimedAttemptSettings prevSettings) {
+      Throwable prevThrowable, ResponseT prevResponse, TimedAttemptSettings prevSettings) {
+    // a small optimization, which allows to avoid calling relatively heavy methods
+    // like timedAlgorithm.createNextAttempt(), when is not necessary.
```
s/when is/when it is/

vam-google replied:
corrected.

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/AttemptCallable.java:
```diff
@@ -0,0 +1,108 @@
+import org.threeten.bp.Duration;
+
+/**
+ * A callable, representing a retriable grpc call. This class is used from {@link RetryingCallable}
```
Corrected

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/AttemptCallable.java:
```diff
@@ -0,0 +1,108 @@
+    return nextContext;
+  }
+
+  private static class NonCancelableFuture<ResponseT> extends AbstractApiFuture<ResponseT> {
```
Corrected

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationCallable.java:
```diff
@@ -108,17 +118,18 @@
+    return futureCall(initialCallable.futureCall(request, context));
+  }
+
+  OperationFuture<ResponseT> futureCall(ApiFuture<Operation> initialFuture) {
```
There is already a futureCall override of this method, which accepts request & context (just above this method). This particular method is package-private. It exists because it is directly used from two other public methods: `futureCall(request, context)` and `resumeFuture()`. This method is not private, mostly because it is called once in test to test some unlikely cancellation/error cases.

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationCallable.java:
```diff
@@ -173,9 +184,65 @@ public ResponseT call(RequestT request) {
+   * @param operationName The name of the operation to cancel.
+   * @return the future which completes once the operation is canceled on the server side.
+   */
+  public ApiFuture<Empty> cancel(String operationName) {
```
OperationFuture was the only heavy future in the library. Basically future is just an object a client can use to wait for completion of something. The actual logic is usually somewhere else. I would like to keep future as simple as possible.The first line of documentation of `java.util.concurrent.Future` says the following:

A Future represents the result of an asynchronous computation.

So a future represents a result of computation, not the computation itself.

Also, for this particular case, cancel() of the server-side operation will conflict with cancel() of the future itself, which are two different things.

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationPollingCallable.java:
```diff
@@ -0,0 +1,125 @@
+  private final OperationsClient operationsClient;
+  private final ApiFuture<Operation> initialFuture;
+
+  OperationPollingCallable(ApiFuture<Operation> initialFuture, OperationsClient operationsClient) {
```
Yes. This class also does a lot of other crazy things (to me at least). It basically, together with AttemptCallable juggles with various futures in brutal manner. 

When I think about it it gives me headache. There are **tasks of different type**, representing some phase of the polling process, all executed by same threadpool, an yet, at any given point in time, there is only **one instance** of a task in the pool (either scheduled or running).

The specific reason, why this class accepts ApiFuture instead of FutureCallable: this allows to pass that future to returned OperationFuture object at the appropriate time. It is sill possible to avoid doing that, but it will involve more settable futures, listeners and weird circular dependencies and interactions between OperationFuture and a OperationPollingCallable. I really-really don't want to do that, especially, when there is a much simpler, more robust and faster solution. Also, this resembles how it was done in previous implementation. Plus this is a package-private class.

Also, originally I had it receiving FutureCallaable initialCallable in the constructor. But ended up changing that to support the OperationFuture case. This was also one of the reasons why I wanted to get rid from OperationFuture at all.

vam-google commented:
PTAL (except the metadata part)

vam-google commented:
PTAL.

1) Addressed feedback.
2) Added execution tracking capabilities in RetryinFuture.
3) Added Metadata parsing support for operation future.
4) Fixed missed polling information issue for operation future (now it is possible to track progress of the operation in client code calling peekMetadata()/getMetadata() methods).
5) Refactored whole retrying future code (split synchronous and asynchronous (callback chaining) parts)
6) Fixed few bugs
7) Added heavyweight multithreaded testing on the retrying logic with maximum coverage and multiple executions of same test (to eliminate sporadic issues).

garrettjonesgoogle commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationPollingCallable.java:
```diff
@@ -99,16 +96,14 @@ public TimedAttemptSettings createNextAttempt(
+      //      Status status = Status.fromCodeValue(prevResponse.getError().getCode());
+      //      if (Status.Code.CANCELLED.equals(status.getCode())) {
+      //        throw new CancellationException();
+      //      }
```
is this dead code that can be deleted?

vam-google replied:
Yes, it is, deleting it. This also implies that in case if client tries to poll an operation which is cancelled on the server side it will get an ApiException (instead of CancellationException) from OperationFuture.get(). This was done on purpose, since ApiException provides information about what actually happened (server canceled operation), while CancellationException does not allow the client to distinguish between him trying to poll a dead operation and him cancelling its own client-side future only. 

As result, OperationFuture cancellation now mean only client-side cancellation, the server side cancellation will result in OperationFuture failure, with ApiException.getCode() == Code.CANCELLED.

garrettjonesgoogle commented on gax-grpc/src/test/java/com/google/api/gax/grpc/CancellationHelpers.java:
```diff
@@ -47,7 +47,9 @@ public static void cancelInThreadAfterLatchCountDown(
-                  while (!resultFuture.cancel(true)) {
+                  System.out.println("Cancelling");
+                  while (!resultFuture.cancel(true) && !resultFuture.isDone()) {
+                    System.out.println("Cancelling again");
```
If we are going to keep these print statements, maybe make this more informative (e.g. why is it cancelling again?)

vam-google replied:
Removed the statements. 
As for why cancelling again: any future (represented by a specific future task, for example `com.google.common.util.concurrent.ListenableFutureTask`, which is the most common in our code), when cancel() is called may or may return false, meaning that cancellation failed. 

Documentation says that most probably it is because the future is already completed/cancelled, but `com.google.common.util.concurrent.ListenableFutureTask` sometimes return false even for the still running futures. I believe (but I is not necessarily the case) this happens because of the rare condidtion, when same FutureTask is being set and cancelled at the same time from different threads, and since Future.set() and Future.cancel() use non-blocking algorithms, if race condition happen they just return `false` instead of blocking and waiting for the thing to proceed "properly".

This becomes even more common case if we are running the AttemptCallable, which explicitly blocks cancellation for a short amount of time setting retryingFuture attempt future to NonCancellableFuture.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/retrying/BasicRetryingFuture.java:
```diff
@@ -0,0 +1,216 @@
+    // since listener is guaranteed to be called only after this future is
+    // completed and this future is guaranteed to be completed only after it is initialized. Also
+    // since "super" is called explicitly here there are no unexpected overrides of addListener here.
+    super.addListener(new CompletionListener(), MoreExecutors.directExecutor());
```
The standard way around this micro crime is to use a static creation method and add listeners after the object is constructed.

vam-google replied:
Yes, that will work, but here that class is extended, moving that logic into factory method will require all children follow same convention and/or avoid doing it. I would like to prevent that from happening.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/retrying/BasicRetryingFuture.java:
```diff
@@ -0,0 +1,216 @@
+
+  @Override
+  // "super." is used here to avoid infinite loops of callback chains
+  public void setAttempt(Throwable throwable, ResponseT response) {
```
The function name doesn't seem to match the intent - maybe `setAttemptResult`?

vam-google replied:
I did not call it setAttemptResult becaus it will interfere with getAttemptResult which, I thought, was a different thing. But now, when I think about it, this method is literally a legit setAttemptResult method, and getAttemptResult/setAttemptResult make a valid setter/getter pair =). So yes, I'll rename it.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/retrying/BasicRetryingFuture.java:
```diff
@@ -0,0 +1,216 @@
+      // Usually should not happen but is still possible, for example if one of the attempt callbacks
+      // throws an exception. An example of such condition is the OperationFuture
+      // which uses ApiFutures.transform(), which actually assign callbacks to the attempt result
+      // futures, and those can fail, for example if metadata class is a wrong one.
```
Add a note about why the exception is swallowed

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/retrying/BasicRetryingFuture.java:
```diff
@@ -0,0 +1,216 @@
+    public void run() {
+      synchronized (lock) {
+        try {
+          //setAttemptFuture(null); TODO: prove that this call is not required
```
Still TODO?

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/retrying/BasicRetryingFuture.java:
```diff
@@ -0,0 +1,216 @@
+ * For internal use only.
+ *
+ * <p>Basic implementation of {@link RetryingFuture} interface. Suitable for usage in "busy loop"
+ * retry implementations.
```
Since this can be subclasses, this last statement could be misleading. Maybe say: "On its own, suitable for usage in 'busy loop' retry implementations. Can be used as the basis for more advanced implementations that do callback chaining."

vam-google replied:
Updated documentation as your suggested (without the last part about callbacks, to not limit the creativity of future extenders =).

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/retrying/NonCancellableFuture.java:
```diff
@@ -0,0 +1,43 @@
+    return false;
+  }
+
+  void cancelPrivately() {
```
So this is a non-cancellable future that can be canceled? :-)

vam-google replied:
Privately =). Well, I could not find a better name for it.

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/AttemptCallable.java:
```diff
@@ -78,7 +77,6 @@ public ResponseT call() {
       externalFuture.setAttemptFuture(internalFuture);
     } catch (Throwable e) {
       externalFuture.setAttemptFuture(ApiFutures.<ResponseT>immediateFailedFuture(e));
-      throw e;
```
Inspired by the recent "learning on the loo" post, suggesting that same exception should not be logged twice. Setting catched exception into externalFuture will record that exception in another place already, so throwing it from here will duplicate same exception.

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationCallable.java:
```diff
@@ -217,31 +223,70 @@ public ResponseT apply(Operation input) {
+      Status status =
+          input.getError() != null
+              ? Status.fromCodeValue(input.getError().getCode())
+              : Status.UNKNOWN;
```
You are right, moved.

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationFuture.java:
```diff
@@ -29,58 +29,117 @@
+  @Override
+  public String toString() {
+    return pollingFuture.toString();
+  }
```
Oops, this is a debugging artifact which I forgot to remove. Removed completely.

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationFuture.java:
```diff
@@ -29,58 +29,117 @@
+
+  private volatile ApiFuture<Operation> peekedAttemptResult;
+  private volatile ApiFuture<MetadataT> peekedPollResult;
+  private volatile ApiFuture<Operation> gottenAttemptResult;
```
Probably, but the naming is inspired by the naming of the method the object is returned from: peeked returned from peek() method, gotten returned from get() method.

vam-google commented:
Addressed review feedback, added few more tests, added documentation for OperationFuture.
PTAL.

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/retrying/RetryingFuture.java:
```diff
@@ -34,27 +34,108 @@
+   *     instead
+   * @param response response returned by the attempt, {@code null} can be a valid response value
+   */
+  void setAttempt(Throwable throwable, ResponseT response);
```
Still to rename to `setAttemptResult`?

vam-google replied:
Please no. AttemptResult is a different thing, basically setAttempt calls setAttemptResult inside it (via same thread callback, or directly on BasicRetryingFuture.java#107

garrettjonesgoogle commented on gax/src/test/java/com/google/api/gax/retrying/ScheduledRetryingExecutorTest.java:
```diff
@@ -47,74 +56,251 @@
+
+  @Test
+  public void testCancelProxiedFutureAfterStart() throws Exception {
+    // this is a heavy test, which takes a lot of time, so only few executions.
```
How long does it take?

vam-google replied:
Just around one second per iteration on my machine, so currently 2 seconds for the test (since there are only 2 iterations). Before commiting I usually check that test with 1000 iterations per test (and yet somehow they manage to fail on travis =(, probably because travis build machine is too much slower than google workstations).

garrettjonesgoogle commented on gax/src/main/java/com/google/api/gax/retrying/BasicResultRetryAlgorithm.java:
```diff
@@ -0,0 +1,68 @@
+ * @param <ResponseT> attempt response type
+ */
+@BetaApi
+public class BasicResultRetryAlgorithm<ResponseT> implements ResultRetryAlgorithm<ResponseT> {
```
Maybe `BaseResultRetryAlgorithm` might be the best match. (Optional)

vam-google replied:
I thought that with respect to the prefixes Base was more for abstract classes (which still have something missing to be fully functional), while basic is for non-abstract (providing minimum workable solution out of the box).

garrettjonesgoogle replied:
I would say you are correct in the general case, but this is used as a base class. The name is potentially confusing when working with the is-a meaning of inheritance, because derived classes are no longer "basic", even though a derived class "is-a" `BasicResultRetryAlgorithm`.

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationFuture.java:
```diff
@@ -40,19 +40,16 @@
 
-/** An ApiFuture which polls a service through OperationsApi for the completion of an operation. */
+/**
+ * An ApiFuture which tracks polling of a service through OperationsApi for the completion of an
```
Provided more up to date and descriptive implementation.

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationFuture.java:
```diff
@@ -65,6 +62,17 @@ public String toString() {
+   *
+   * @param pollingFuture retrying future which tracks polling of the server operation (in most
+   *     cases with exponential upper bounded intervals)
+   * @param initialFuture the initial future which started operation on the server side
```
Corrected

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationFuture.java:
```diff
@@ -131,7 +175,34 @@ public String getName() throws ExecutionException, InterruptedException {
+   * ApiFuture#get()} is a potentially blocking operation. To get metadata without blocking the
+   * current thread use the {@link #peekMetadata()} method instead.
+   *
+   * <p>In general this method behaves similarly to {@link RetryingFuture#getAttemptResult()}.
```
Agree. I moved that line at the end of the documentation block, so only a very meticulous reader will ever read that line =). I still want to keep it there to provide a deeper look on how the method behaves. Even though conceptually it is easy to understand what the method is about, the way how it behaves has many pitfalls. There are many corner cases, possible outcomes and execution ordering combinations. It would be nice if a user, who got "weird" behavior, to get the explanation directly from the documentation without digging too much into implementation details. 

The main purpose of the link to RetryingFuture here is to document how and in which order attempts/attempt results/the whole future are resolved.

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/AttemptCallable.java:
```diff
@@ -0,0 +1,99 @@
+import org.threeten.bp.Duration;
+
+/**
+ * A callable, representing a retriable grpc call. This class is used from {@link RetryingCallable}.
```
Removed

vam-google commented on gax-grpc/src/main/java/com/google/api/gax/grpc/OperationCallable.java:
```diff
@@ -29,70 +29,86 @@
   private final OperationsClient operationsClient;
-  private final Class<ResponseT> responseClass;
-  private final OperationCallSettings settings;
+  private final ApiFunction<Operation, ResponseT> resultTransformer;
```
Modified here and in the OperationFuture

vam-google commented:
PTAL

garrettjonesgoogle commented:
The changes LGTM
